pubsub/acks_test.go

```diff
@@ -342,3 +342,86 @@ func TestReceiveReturnsErrorFromSendAcks(t *testing.T) {
 		}
 	}
 }
+
+// callbackDriverSub implements driver.Subscription and allows something like
+// monkey patching of both its ReceiveBatch and SendAcks methods.
+type callbackDriverSub struct {
+	driver.Subscription
+	receiveBatch func(context.Context) ([]*driver.Message, error)
+	sendAcks     func(context.Context, []driver.AckID) error
+}
+
+func (s *callbackDriverSub) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	return s.receiveBatch(ctx)
+}
+
+func (s *callbackDriverSub) SendAcks(ctx context.Context, acks []driver.AckID) error {
+	return s.sendAcks(ctx, acks)
+}
+
+func (s *callbackDriverSub) IsRetryable(error) bool { return false }
+
+func (s *callbackDriverSub) ErrorCode(error) gcerrors.ErrorCode { return gcerrors.Internal }
+
+// This test detects the root cause of
+// https://github.com/google/go-cloud/issues/1238.
+// If the issue is present, this test times out.
+func TestReceiveReturnsAckErrorOnNoMoreMessages(t *testing.T) {
+	// If SendAcks fails, the error is returned via receive.
+	ctx := context.Background()
+	serr := errors.New("unrecoverable error")
+	receiveHappened := make(chan struct{})
+	ackHappened := make(chan struct{})
+	ds := &callbackDriverSub{
+		// First call to receiveBatch will return a single message.
+		receiveBatch: func(context.Context) ([]*driver.Message, error) {
+			ms := []*driver.Message{{AckID: 1}}
+			return ms, nil
```

Comment:
You don't need the variable.
```
return []*driver......{}, nil
```

Reply:
That's true. I find it slightly more readable the way I wrote it though.

Reply:
@zombiezen @vangent @shantuo Is anyone else bothered by these one-line temporary variables? I perceive them like little bumps in the road, not terrible but kind of annoying. If I'm the only one then I'll shut up.

Reply:
It did cause me pause while reading in the way you describe, @jba. Unfortunately, I can't quite vocalize why in the general case. In this particular case, this caused me to pause because I thought it was mutating a variable, but then I had to read the colon in the `:=` operator.

Reply:
In that case I suppose I could write it as
```
var ms = []*driver.Message{{AckID: 1}}
return ms, nil
```

Reply:
Please don't do that.

Reply:
Too verbose?

Reply:
Do I have to inline this to get the PR approved?

Reply:
No, sorry, I was just waiting for people to weigh in. 

I came back to approve, but please change the `var ms =` back to `ms :=`.

Reply:
fwiw, it does not pain me to read one-line temporary variables like this. I'm kinda on the side of "it's more readable this way", bc it seems like the `nil` could get lost at the end of `return []*driver.Message{{AckID: 1}}, nil`.

Reply:
That's why I often do it this way.